### PR TITLE
allow handler to join after the informer has started

### PR DESCRIPTION
**Pull request by deads2k**
This allows an event handler to join after a SharedInformer has started.  It can't add any indexes, but it can add its reaction functions.

This works by 
1. stopping the flow of events from the reflector (thus stopping updates to our store)
2. registering the new handler
3. sending synthetic "add" events to the new handler only
4. unblocking the flow of events

It would be possible to 
1. block
2. list
3. add recorder
4. unblock
5. play list to as-yet unregistered handler
6. block
7. remove recorder
8. play recording
9. add new handler
10. unblock

But that is considerably more complicated.  I'd rather not start there since this ought to be the exception rather than the rule.

@wojtek-t who requested this power in the initial review
@smarterclayton @liggitt I think this resolves our all-in-one ordering problem.
@hongchaodeng since this came up on the call

### CODE REVIEW

**Comment by wojtek-t** [comment]
@deads2k - great thanks!
I'm reducing my backlog after vacation last week. Will take a look probably tomorrow.

File: pkg/controller/framework/shared_informer.go [hunk 1]
```diff
@@ -199,12 +203,28 @@ func (s *sharedIndexInformer) AddEventHandler(handler ResourceEventHandler) erro
...
+		return nil
 	}
 
+	// in order to safely join, we have to
+	// 1. stop sending add/update/delete notifications
+	// 2. do a list against the store
+	// 3. send synthetic "Add" events to the new handler
+	// 4. unblock
+	s.blockDeltas.Lock()
+	defer s.blockDeltas.Unlock()
```
Comment (wojtek-t) [inline pkg/controller/framework/shared_informer.go]: This one is OK, but blockDeltas should also block sending deltas, which is currently doesn't.

**Comment by wojtek-t** [comment]
@deads2k - I added one comment

File: pkg/controller/framework/shared_informer.go (hunk 1 above)
Comment (deads2k) [inline pkg/controller/framework/shared_informer.go]: > This one is OK, but blockDeltas should also block sending deltas, which is currently doesn't.

Fixed.  Bad port from downstream :(

**Comment by wojtek-t** [comment]
Thanks - LGTM

**Pull request merged**